Drain services in Sidecar

mihaitodor opened the pull request:
This contains the required changes to trigger the new `DRAINING` state [implemented](https://github.com/Nitro/sidecar/pull/47) in Sidecar. It also contains some small fixes and adds a bunch of tests.

In detail:
- 504fe84 - Stop trying to fetch state from Sidecar when successful: Unless I'm mistaking, the code was fetching the state from Sidecar `SidecarRetryCount` times in each iteration of the `watchLooper`, even if the Sidecar call would succeed. I believe I have fixed this.
- 827210c - Turn `StatusSleepTime` into a parameter of `sidecarExecutor`, so the code is easier to test.
- b6765e4 - Check if the Mesos driver is running before attempting to stop it: Trying to stop the Mesos driver when it's status is not `DRIVER_RUNNING` [causes it to return an error](https://github.com/Nitro/sidecar-executor/blob/79f079f6e342963c8dc595c7d4f1f3f582fa0749/vendor/github.com/mesos/mesos-go/api/v0/executor/executor.go#L573). This issue has surfaced after #43.
- 0709064 - Make Mesos Master port configurable: Enables more testing.
- c4734c2 - Drain services in Sidecar before stopping them: This contains the actual implementation for triggering the new `DRAINING` state in Sidecar.
    - I cached the `containerConfig` and the `containerID` on the `sidecarExecutor` object because I need to reuse them in `KillTask` and in `drainService`.
    - Also, I'm reusing the `SidecarUrl` which defaults to `http://localhost:7777/state.json` for figuring out the URL to which I need to send the drain service POST request.
    - cb06719 - Stop waiting for the service to drain if the container exits prematurely
- a9bec29 - Cleanup global config: Having `config` as a global object makes testing very difficult. I have refactored the code to always read the config from `sidecarExecutor`.
- 2fa11b9 - Add loads of tests for the callbacks.
- f4a4777 - Update Readme to document SidecarDrainingDuration and MesosMasterPort.

mihaitodor commented on callbacks.go:
```diff
@@ -150,14 +170,9 @@ func (exec *sidecarExecutor) KillTask(driver executor.ExecutorDriver, taskID *me
 	exec.sendStatus(status, taskID)
 
-	// We have to give the driver time to send the message
-	time.Sleep(StatusSleepTime)
```
FIY: I moved this into `exec.sendStatus`. We need to sleep each time we send a status. Maybe the new version of the Mesos API does this better.

relistan replied:
We don't need to sleep every time we send a status. We only have to sleep when sending a status _at_ _shutdown_.

relistan commented:
> OK, fair enough. I was thinking this will offer more flexibility, since the user can configure the time per service. I'm OK to drop it.

They can configure it per service. This is done in the `executor` section of the `nmesos` config.

relistan commented:
This is super good work, @mihaitodor ! A few comments to fix, but nothing big. 🎉

mihaitodor commented:
Hey @relistan, I have addressed your comments and updated the following commits:
- b6765e4 - Check if the Mesos driver is running before attempting to stop it: I reverted the `time.Sleep` changes to `sendStatus`
- c4734c2 - Drain services in Sidecar before stopping them: renamed `drainService` to `notifyDrain`, used an URL to create the Sidecar `drain` URL and removed the useless code for processing the `SidecarDrainingDuration` label, since we can configure it via the `EXECUTOR_SIDECAR_DRAINING_DURATION` env var, as you mentioned.
- cb06719 - I made an attempt to address [your comment](https://github.com/Nitro/sidecar-executor/commit/51902fc676f65c07bbb47d9787636a50c5f14fc1#diff-a4b28ce65f43b814b9b071c40b254382R223) on stopping the `DRAINING` sleep early if the container gets killed in the mean time. I wanted to avoid adding an extra channel on the executor, but, unfortunately, I can't just use `exec.watchLooper.Wait()` because we're already waiting on it in [`monitorTask`](https://github.com/Nitro/sidecar-executor/blob/79f079f6e342963c8dc595c7d4f1f3f582fa0749/executor_utils.go#L29)

relistan commented:
:shipit:
